plugins/thunderbolt/fu-thunderbolt-image-validation.c

```diff
@@ -0,0 +1,482 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "fu-thunderbolt-image-validation.h"
+
+#include <string.h>
+#include <libfwupd/fwupd-error.h>
+
+enum Sections {
+	DIGITAL_SECTION,
+	DROM_SECTION,
+	ARC_PARAMS_SECTION,
+	DRAM_UCODE_SECTION,
+	SECTION_COUNT
+};
+
+typedef struct {
+	enum Sections section; /* default is DIGITAL_SECTION */
+	guint32 offset;
+	guint32 len;
+	guint8 mask;           /* 0 means "no mask" */
+	gchar* description;
+} FwLocation;
+
+typedef struct {
+	const guint8 *data;
+	gsize len;
+	guint32 *sections;
+} FwObject;
+
+typedef struct {
+	guint16 id;
+	guint gen;
+	guint ports;
+} HwInfo;
+
+static HwInfo *
+get_hw_info (guint16 id)
+{
+	static const HwInfo hw_info_arr[] = {
+		{ 0x156D, 2, 2 }, /* FR 4C */
+		{ 0x156B, 2, 1 }, /* FR 2C */
+		{ 0x157E, 2, 1 }, /* WR */
+
+		{ 0x1578, 3, 2 }, /* AR 4C */
+		{ 0x1576, 3, 1 }, /* AR 2C */
+		{ 0x15C0, 3, 1 }, /* AR LP */
+		{ 0x15D3, 3, 2 }, /* AR-C 4C */
+		{ 0x15DA, 3, 1 }, /* AR-C 2C */
+
+		{ 0 }
+	};
+
+	for (int i = 0; hw_info_arr[i].id != 0; i++)
+		if (hw_info_arr[i].id == id)
+			return hw_info_arr + i;
+	return NULL;
+}
+
+static inline gboolean
+valid_farb_pointer (guint32 pointer)
+{
+	return pointer != 0 && pointer != 0xFFFFFF;
+}
+
+static GByteArray *
+read_location (const FwLocation *location, const FwObject *fw)
+{
+	guint32 location_start = fw->sections[location->section] + location->offset;
+	g_autoptr(GByteArray) read = g_byte_array_new ();
+
+	// TODO: set error?
+	if (location_start > fw->len || location_start + location->len > fw->len)
+		return NULL;
+
+	read = g_byte_array_append (read,
+				    fw->data + location_start,
+				    location->len);
+
+	if (location->mask)
+		read->data[0] &= location->mask;
+	return g_steal_pointer (&read);
+}
+
+static guint32
+read_farb_pointer_impl (const FwLocation *location, const FwObject *fw)
+{
+	g_autoptr(GByteArray) farb = read_location (location, fw);
+	farb = g_byte_array_set_size (farb, sizeof(guint32));
+	memset (farb->data + location->len, 0, farb->len - location->len);
+	return GUINT32_FROM_LE (*(guint32*) farb->data);
+}
+
+static guint32
+read_farb_pointer (const FwObject *fw)
+{
+	const FwLocation farb1 = { .offset = 0,      .len = 3 };
+	const FwLocation farb2 = { .offset = 0x1000, .len = 3 };
+
+	guint32 ret = read_farb_pointer_impl (&farb1, fw);
+	if (valid_farb_pointer (ret))
+		return ret;
+	return read_farb_pointer_impl (&farb2, fw);
+}
+
+static gboolean
+compare (const FwLocation *location, const FwObject *controller_fw, const FwObject *image_fw)
+{
+	g_autoptr(GByteArray) controller_data = read_location (location, controller_fw);
+	g_autoptr(GByteArray) image_data      = read_location (location, image_fw);
+
+	if (!image_data || !controller_data ||
```

Comment:
In fwupd I'd much prefer `image_data == NULL` rather than `!image_data` just to clarify it's not a boolean. Would `g_return_val_if_fail` be more useful here? Otherwise same comment about GError requirement.

Reply:
Style - sure. Fixing.
`g_return_val_if_fail` - depends on the importance of the check. I see from the documentation that such a check can be disabled. Do we want it to be disabled under some conditions? The result of using a wrong image file (supplied locally y user, for example) would be null dereference or reading outside of the buffer boundaries.